File: server/express/framework/Service.ts
import InstallStatus from "express/models/InstallStatus"
import Package from "express/models/Package"
import path from "path"
import Main from "../../electron/ElectronStarter"
import Gateway from "../interfaces/Gateway"
import Message from "../models/Message"
import Status from "../models/Status"
import { SubscriptionListener } from "../models/SubscriptionListener"
import RobotLabXRuntime from "../service/RobotLabXRuntime"
import { CodecUtil } from "./CodecUtil"
import { getLogger } from "./Log"

const log = getLogger("Service")

interface NotifyList {
  [key: string]: SubscriptionListener[]
}

export default class Service implements Gateway {
  protected startTime: number | null = null

  id: string | null = null
  name: string | null = null
  typeKey: string | null = null
  version: string | null = null
  hostname: string | null = null
  fullname: string | null = null

  // User "owned" data path for resources for the service
  dataPath: string | null = null

  notifyList: NotifyList = {}

  pkg: Package | null = null

  ready: boolean = false

  installed: boolean = false

  config: any = {}

  constructor(id: string, name: string, typeKey: string, version: string, hostname: string | null = null) {
    this.id = id
    this.name = name
    this.typeKey = typeKey
    this.version = version
    this.hostname = hostname
    this.fullname = `${this.name}@${this.id}`
    // FIXME should be expressRoot/data
    this.dataPath = path.join(Main.expressRoot, `service/${this.name}`)
  }

  getSubscribersForMethod(method: string): SubscriptionListener[] {
    let ret: string[] = []
    const listeners: SubscriptionListener[] = this.notifyList[method]
    if (listeners) {
      listeners.forEach((listener: SubscriptionListener) => {
        ret.push(listener.callbackName)
      })
    }
    log.error(`getSubscribersForMethod ${method} ${ret}`)
    return []
  }

  addListener(method: string, remoteName: string, remoteMethod: string) {
    // ensure remoteName is a fullname
    if (!remoteName.includes("@")) {
      remoteName = CodecUtil.getFullName(remoteName)
    }

    if (remoteMethod === null || remoteMethod === "" || remoteMethod === undefined) {
      remoteMethod = CodecUtil.getCallbackTopicName(method)
    }

    // log.info(`== addListener ${this.name}.${method} --> ${remoteName}.${remoteMethod}`)

    if (!(method in this.notifyList)) {
      this.notifyList[method] = []
    }
    const listeners = this.notifyList[method] || []
    for (const listener of listeners) {
      if (listener.callbackName === remoteName && listener.callbackMethod === remoteMethod) {
        // log.info(`listener on ${method} for -> ${remoteName}.${remoteMethod} already exists`)
        return listener
      }
    }

    const listener = new SubscriptionListener(method, remoteName, remoteMethod)
    this.notifyList[method].push(listener)
    return listener
  }

  /**
   * Broadcasts the current state of the service
   * @returns returns self
   */
  broadcastState() {
    return this
  }

  getConfig(): any {
    return this.config
  }

  applyConfig(config: any) {
    this.config = config
  }

  apply(config: any) {
    this.applyConfig(config)
  }

  // TODO - handle ad hoc config file
  applyFileConfig(filename: string = null) {
    RobotLabXRuntime.getInstance().applyServiceFileConfig(this.name)
    this.invoke("broadcastState")
  }

  saveConfig() {
    RobotLabXRuntime.getInstance().saveServiceConfig(this.name, this.config)
  }

  getNotifyList() {
    return this.notifyList
  }

  getHostname(): string | null {
    return this.hostname
  }

  getId() {
    return this.id
  }

  getName() {
    return this.name
  }

  // Example of calculating uptime
  getUptime(): string {
    if (!this.startTime) {
      return "service not started"
    }
    const now = new Date()
    const uptime = now.getTime() - this.startTime
    return `uptime: ${uptime / 1000} seconds`
  }

  invoke(methodName: string, ...args: any[]): any {
    let msg = new Message(this.name, methodName, args)
    msg.sender = this.fullname
    return this.invokeMsg(msg)
  }

  invokeMsg(msg: Message): any {
    try {
      const fullName = this.fullname //this.getFullName()
      const msgFullName = CodecUtil.getFullName(msg.name)
      const msgId = CodecUtil.getId(msgFullName)
      const senderId = CodecUtil.getId(msg.sender)
      const runtime = RobotLabXRuntime.getInstance()
      // this process's id
      const id = runtime.getId()
      let ret: any = null

      if (msg.data && msg.data.length > 0) {
        // log.info(`--> ${msg.sender} --> ${msg.name}.${msg.method}(${JSON.stringify(msg.data)})`)
        log.info(`--> ${msg.sender} --> ${msg.name}.${msg.method}(...)`)
      } else {
        log.info(`--> ${msg.sender} --> ${msg.name}.${msg.method}()`)
      }

      // FIXME - building dynamic routes based on "registration"
      // however this filter could be opened up to any id from sender that is not the same as the current id
      // It should be done this way, and all external service connections should be registered with
      // generated uuids, for sender process remoteIds

      // log.error(`invokeMsg msgId ${msgId} id ${id} msgFullName ${msgFullName} fullName ${fullName}`)
      // ==== REMOTE ====
      // FIXME - check if blocking or non-blocking
      // use gateway to send message to remote service
      if (msgId !== id) {
        // log.info(`remote message ${msgFullName}.${msg.method} from ${msg.sender}.${msg.method}`)
        // dynamically add route to gateway
        if (msg.gateway) {
          runtime.addRoute(msgId, msg.gatewayId, msg.gateway)
        }

        // send message to remote service
        // log.info(`sending message to ${msgFullName}.${msg.method}`)
        // this.gateway.send(msg)
        // const json = JSON.stringify(msg)
        log.info(`<-- ${msgFullName}.${msg.method} <-- ${msg.sender}.${msg.method}`)
        // FIXME bork'd - need state information regarding connectivity of process/service, and its an "array" of connections
        log.info(`connectionImpl / connections ${[...runtime.getClients().keys()]} `)

        // fine the gateway for the message's remoteId
        let gateway: Gateway = runtime.getGateway(msgId)
        if (!gateway) {
          log.error(`NO GATEWAY for remoteId ${msgId}`)
          return null
        }

        // TODO - implement synchronous blocking
        let blockingObject = gateway.sendRemote(msg)

        // TODO - implement synchronous blocking
        return blockingObject
      }

      // ==== LOCAL PROCESS DIFFERENT SERVICE ====
      // get the service - asynchronous buffered or synchronous non-buffered
      // default synchronous non-buffered
      if (msgFullName !== fullName) {
        log.info(
          `message to different service ${msgFullName} !== ${fullName} ${msgFullName}.${msg.method} from ${msg.sender}.${msg.method}`
        )
        let service = runtime.getService(msgFullName)
        if (service === null) {
          log.error(`service ${msgFullName} not found`)
          return null
        } else {
          // relay to correct service
          // service.send(msg)
          return service.invokeMsg(msg)
        }
      }

      // ==== LOCAL ====
      // FIXME - check if blocking or non-blocking
      // is this the service to invoke the method on ?
      if (fullName === msgFullName) {
        log.info(`(invoke) ${msgFullName}.${msg.method} from ${msg.sender}.${msg.method}`)
        let obj: any = this // cast away typescript

        if (!msg.method) {
          // ui error - user should be informed
          console.error(`method ${msg.method} not found`)
          return null
        }

        // invoke locally
        log.debug(`invoking ${this.name}.${msg.method}`)
        try {
          if (msg.data && msg.data.length > 0) {
            ret = obj[msg.method](...msg.data)
          } else {
            ret = obj[msg.method]()
          }
        } catch (e) {
          log.error(`failed to invoke ${this.name}.${msg.method} because ${e}`)
        }

        // normalize undefined to null
        if (ret === undefined) {
          ret = null
        }

        // TODO - process subscription
        if (this.notifyList[msg.method]) {
          this.notifyList[msg.method].forEach((listener: any) => {
            let subMsg = new Message(listener.callbackName, listener.callbackMethod, [ret])
            subMsg.sender = this.fullname
            // log.info(`<- notify ${listener.callbackName}.${listener.callbackMethod}`)
            this.invokeMsg(subMsg)
          })
        }
        return ret
      }
    } catch (e) {
      // ui error - user should be informed
      log.error(`failed to invoke ${this.name}.${msg.method} because ${e}`)
      if (e instanceof Error) {
        log.error(e.stack)
      } else {
        log.error("Caught an unknown error type:", e)
      }
    }
    return null
  }

  isReady(): boolean {
    return this.ready
  }

  publishStdOut(msg: string): string {
    log.info(`stdout: ${msg}`)
    return msg
  }

  // maybe purge is a good idea to purge the copy of the repo for an instance ?
  // vs release which only frees the service from memory
  releaseService() {
    log.info(`========= released service ${this.getName()} ===========`)
    RobotLabXRuntime.getInstance().release(this.fullname)
  }

  removeListener(method: string, remoteName: string, remoteMethod: string) {
    // log.warn(`removeListener ${method} ${remoteName} ${remoteMethod}`)

    if (remoteMethod === null || remoteMethod === undefined) {
      // log.info(`remoteMethod is null, setting to ${CodecUtil.getCallbackTopicName(method)}`)
      remoteMethod = CodecUtil.getCallbackTopicName(method)
    }

    // log.info(`== removeListener ${this.name}.${method} --> ${remoteName}.${remoteMethod}`)

    if (!this.notifyList || !this.notifyList.hasOwnProperty(method)) {
      // log.error("no listeners for method " + method)
      return
    }

    this.notifyList[method].forEach((listener: any, index: any) => {
      // log.info(
      //   `checking listener ${listener.callbackName}.${listener.callbackMethod} for ${remoteName}.${remoteMethod}`
      // )
      if (listener.callbackName === remoteName && listener.callbackMethod === remoteMethod) {
        this.notifyList[method].splice(index, 1)
        //        log.info(`removed listener on ${method} for -> ${remoteName}.${remoteMethod}`)
        return
      }
    })

    //log.error("no listeners for method " + method)
  }

  publishStatus(status: Status) {
    if (status.level === "error") {
      log.error(status.detail)
    } else if (status.level === "warn") {
      log.warn(status.detail)
    } else {
      log.info(status.detail)
    }
    return status
  }

  /**
   * A generalized install status to determine if dependencies have
   * successfully been installed, and details of problems.
   *
   * @param status
   * @returns
   */
  publishInstallStatus(status: InstallStatus) {
    return status
  }

  // Example of a shared method
  startService() {
    this.startTime = new Date().getTime()
    this.ready = true
    log.info(`========= started service ${this.name} ===========`)
  }

  stopService() {
    this.startTime = null
    this.ready = false
    log.info(`========= stopped service ${this.name} ===========`)
  }

  info(msg: string | null) {
    this.invoke("publishStatus", new Status("info", msg, this.name))
  }

  warn(msg: string | null) {
    this.invoke("publishStatus", new Status("warn", msg, this.name))
  }

  error(msg: string | null) {
    this.invoke("publishStatus", new Status("error", msg, this.name))
  }

  /**
   * Requesting to send a message to a remote process
   * @param msg
   */
  public sendRemote(msg: Message): void {
    // default is runtime's sendRemote
    RobotLabXRuntime.getInstance().sendRemote(msg)
  }

  public setInstalled(installed: boolean) {
    this.installed = installed
  }

  public save() {
    this.saveConfig()
  }

  toJSON() {
    return {
      config: this.config,
      fullname: this.fullname,
      hostname: this.hostname,
      id: this.id,
      installed: this.installed,
      name: this.name,
      notifyList: this.notifyList,
      pkg: this.pkg,
      ready: this.ready,
      typeKey: this.typeKey,
      version: this.version,
      startTime: this.startTime
    }
  }
}

File: python-server/app/models/service.py
from typing import Optional, Dict, List, Any
from pydantic import BaseModel, model_validator, Field
import os
import logging
from .message import Message
from utils.lang_util import convert_to_snake

logger = logging.getLogger("Service")

# Placeholder for SubscriptionListener, Status, InstallStatus, Package, etc.
class SubscriptionListener:
    def __init__(self, method: str, callback_name: str, callback_method: str):
        self.method = method
        self.callbackName = callback_name
        self.callbackMethod = callback_method

class Status:
    def __init__(self, level: str, detail: str, name: Optional[str] = None):
        self.level = level
        self.detail = detail
        self.name = name

class InstallStatus:
    def __init__(self, status: str, detail: str):
        self.status = status
        self.detail = detail

class Service(BaseModel):
    # allow arbitrary types if needed
    model_config = {"arbitrary_types_allowed": True}

    # Field declarations
    id:            Optional[str]        = None
    name:          Optional[str]        = None
    type_key:      Optional[str]        = None
    version:       Optional[str]        = None
    hostname:      Optional[str]        = None
    public_root:   Optional[str]        = None
    start_time:    Optional[float]      = None

    fullname:      Optional[str]        = None
    data_path:     Optional[str]        = None
    notify_list:   Dict[str, List[Any]] = Field(default_factory=dict)
    pkg:           Optional[Any]        = None
    config:        Dict[str, Any]       = Field(default_factory=dict)
    ready:         bool                 = False
    installed:     bool                 = False

    @model_validator(mode="before")
    def build_derived(cls, values: Dict[str, Any]) -> Dict[str, Any]:
        name = values.get("name")
        _id = values.get("id")
        if name and _id:
            values.setdefault("fullname", f"{name}@{_id}")
        public_root = values.get("public_root")
        if public_root and name:
            values.setdefault(
                "data_path",
                os.path.join(public_root, f"service/{name}")
            )
        return values

    def get_subscribers_for_method(self, method: str) -> List[str]:
        listeners = self.notify_list.get(method, [])
        ret = [listener.callbackName for listener in listeners]
        logger.error(f"getSubscribersForMethod {method} {ret}")
        return []

    def send_to(self, name: str, method: str, data: List[Any]):
        logger.info(f"sendTo {name} {method} {data}")
        msg = Message(name=name, method=method, data=data)
        msg.sender = self.fullname
        self.invoke_msg(msg)

    def create_message(self, in_name: str, in_method: str, in_params: List[Any]):
        id_ = self.get_id()
        msg = Message(name=in_name, method=in_method, data=in_params)
        msg.sender = f"runtime@{id_}"
        return msg

    def subscribe_to(self, name: str, method: str):
        logger.info(f"subscribeTo {name} {method}")
        # Placeholder: actual implementation would add to notify_list

    def add_listener(self, method: str, remote_name: str, remote_method: str = None):
        """
        Add a listener to the notify_list for a given method.
        remote_name will be ensured to be a fullname (e.g. name@id).
        remote_method will default to the callback topic name if not provided.
        Mirrors Service.addListener from TypeScript.
        """
        from app.utils.codec_util import CodecUtil
        # Ensure remote_name is a fullname
        if '@' not in remote_name:
            remote_name = CodecUtil.get_full_name(remote_name)
        # Set remote_method to callback topic if not provided
        if not remote_method:
            remote_method = CodecUtil.get_callback_topic_name(method)
        if not hasattr(self, 'notify_list') or self.notify_list is None:
            self.notify_list = {}
        if method not in self.notify_list:
            self.notify_list[method] = []
        # Add or update listener entry
        listener_entry = type('Listener', (), {})()
        listener_entry.callbackName = remote_name
        listener_entry.callbackMethod = remote_method
        self.notify_list[method].append(listener_entry)

    def broadcast_state(self):
        return self

    def get_config(self) -> dict:
        return self.config

    def apply_config(self, config: dict):
        logger.info(f"applyConfig {self.name} {config}")
        self.config = config

    def apply_config_value(self, key: str, value: Any):
        self.config[key] = value
        self.invoke("broadcast_state")

    def apply_file_config(self, filename: Optional[str] = None):
        # Placeholder for RobotLabXRuntime logic
        self.invoke("broadcast_state")

    def save_config(self):
        # Placeholder for RobotLabXRuntime logic
        pass

    def get_notify_list(self):
        return self.notify_list

    def get_hostname(self) -> Optional[str]:
        return self.hostname

    def get_id(self):
        return self.id

    def get_methods(self, filters: Optional[List[str]] = None) -> List[str]:
        method_names = [func for func in dir(self)
                        if callable(getattr(self, func)) and not func.startswith("__") and func != "__init__"]
        if filters:
            method_names = [name for name in method_names if any(name.startswith(f) for f in filters)]
        return sorted(method_names)

    def get_name(self):
        return self.name

    def get_uptime(self) -> str:
        import time
        if not self.start_time:
            return "service not started"
        uptime = time.time() - self.start_time
        return f"uptime: {uptime} seconds"

    def invoke(self, method_name: str, *args: Any) -> Any:
        msg = Message(name=self.name, method=method_name, data=list(args))
        msg.sender = self.fullname
        return self.invoke_msg(msg)

    def invoke_msg(self, msg: Message) -> Any:
        """
        Invokes the method on this service as described by msg.method, passing in msg.data as arguments.
        Handles local, remote, and subscription notification logic. Mirrors Service.ts invokeMsg.
        """
        ret = None
        try:
            # Determine fullName and id for this service and message
            my_fullname = getattr(self, "fullname", None)
            my_id = getattr(self, "id", None)
            msg_fullname = getattr(msg, "name", None)
            msg_id = None
            if msg_fullname and "@" in msg_fullname:
                msg_id = msg_fullname.split("@")[-1]

            # REMOTE MESSAGE HANDLING
            # If the message is not for this process, forward it to the correct gateway
            if msg_id and msg_id != my_id:
                from services.py_robotlabx_runtime import PyRobotLabXRuntime
                runtime = PyRobotLabXRuntime.get_instance()
                gateway = None
                if hasattr(runtime, "get_gateway"):
                    gateway = runtime.get_gateway(msg_id)
                else:
                    # fallback: try to get the connection directly
                    gateway = runtime.connections.get(msg_id)
                if not gateway:
                    logger.error(f"NO GATEWAY for remoteId {msg_id}")
                    return None
                if hasattr(gateway, "send_remote"):
                    return gateway.send_remote(msg)
                elif hasattr(gateway, "send"):
                    return gateway.send(msg)
                else:
                    logger.error(f"Gateway for {msg_id} does not support send_remote/send")
                    return None

            # LOCAL MESSAGE HANDLING
            method_name = convert_to_snake(msg.method)
            if hasattr(self, method_name):
                method = getattr(self, method_name)
                args = msg.data if hasattr(msg, 'data') and msg.data else []
                try:
                    ret = method(*args)
                except Exception as e:
                    self.error(f"failed to invoke {self.name}.{msg.method} because {e}")
            else:
                self.error(f"Method {method_name} not found on {self.name}")

            # Notify listeners if any
            if hasattr(self, 'notify_list') and self.notify_list and msg.method in self.notify_list:
                for listener in self.notify_list[msg.method]:
                    from app.models.message import Message
                    sub_msg = Message()
                    sub_msg.name = listener.callbackName
                    sub_msg.method = listener.callbackMethod
                    sub_msg.data = [ret]
                    sub_msg.sender = getattr(self, 'fullname', None)
                    self.invoke_msg(sub_msg)
            return ret
        except Exception as e:
            logger.error(f"general catch failed to invoke {self.name}.{msg.method} because {e}")
            return None

    def is_ready(self) -> bool:
        return self.ready

    def publish_stdout(self, msg: str) -> str:
        logger.info(f"stdout: {msg}")
        return msg

    def release_service(self):
        logger.info(f"========= released service {self.get_name()} ===========")
        # Placeholder for RobotLabXRuntime.release

    def remove_listener(self, method: str, remote_name: str, remote_method: str):
        if not self.notify_list or method not in self.notify_list:
            return
        listeners = self.notify_list[method]
        self.notify_list[method] = [l for l in listeners if not (l.callbackName == remote_name and l.callbackMethod == remote_method)]

    def publish_status(self, status: Status):
        if status.level == "error":
            logger.error(status.detail)
        elif status.level == "warn":
            logger.warning(status.detail)
        else:
            logger.info(status.detail)
        return status

    def publish_install_status(self, status: InstallStatus):
        return status

    def start_service(self):
        import time
        self.start_time = time.time()
        self.ready = True
        logger.info(f"========= started service {self.name} ===========")

    def stop_service(self):
        self.start_time = None
        self.ready = False
        logger.info(f"========= stopped service {self.name} ===========")

    def info(self, msg: Optional[str]):
        logger.info(msg)
        self.invoke("publish_status", Status("info", msg or "", self.name))

    def warn(self, msg: Optional[str]):
        logger.warning(msg)
        self.invoke("publish_status", Status("warn", msg or "", self.name))

    def error(self, msg: Optional[str]):
        logger.error(msg)
        self.invoke("publish_status", Status("error", msg or "", self.name))

    def send_remote(self, msg: Message):
        from services.py_robotlabx_runtime import PyRobotLabXRuntime
        PyRobotLabXRuntime.get_instance().send_remote(msg)

    def set_installed(self, installed: bool):
        self.installed = installed

    def save(self):
        self.save_config()

    def to_json(self):
        return {
            "config": getattr(self, "config", {}),
            "fullname": getattr(self, "fullname", None),
            "hostname": getattr(self, "hostname", None),
            "id": getattr(self, "id", None),
            "installed": getattr(self, "installed", False),
            "name": getattr(self, "name", None),
            "notifyList": getattr(self, "notify_list", {}),
            "pkg": getattr(self, "pkg", None),
            "ready": getattr(self, "ready", False),
            "typeKey": getattr(self, "type_key", None),
            "version": getattr(self, "version", None),
            "startTime": getattr(self, "start_time", None)
        }

    def __json__(self):
        return self.to_json()

    def __repr__(self):
        return str(self.to_json())
